samza-sql/src/main/java/org/apache/samza/sql/translator/QueryTranslator.java

```diff
@@ -0,0 +1,112 @@
+*
+* Unless required by applicable law or agreed to in writing,
+* software distributed under the License is distributed on an
+* "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+* KIND, either express or implied.  See the License for the
+* specific language governing permissions and limitations
+* under the License.
+*/
+
+package org.apache.samza.sql.translator;
+
+import java.util.Map;
+import java.util.Optional;
+
+import org.apache.calcite.rel.RelNode;
+import org.apache.calcite.rel.RelRoot;
+import org.apache.calcite.rel.RelShuttleImpl;
+import org.apache.calcite.rel.core.TableScan;
+import org.apache.calcite.rel.logical.LogicalFilter;
+import org.apache.calcite.rel.logical.LogicalProject;
+import org.apache.samza.operators.MessageStreamImpl;
+import org.apache.samza.operators.StreamGraph;
+import org.apache.samza.sql.data.SamzaSqlRelMessage;
+import org.apache.samza.sql.interfaces.RelSchemaConverter;
+import org.apache.samza.sql.interfaces.RelToSamzaMsgConverter;
+import org.apache.samza.sql.interfaces.SamzaToRelMsgConverter;
+import org.apache.samza.sql.interfaces.SourceSchemaResolver;
+import org.apache.samza.sql.interfaces.SystemStreamConfig;
+import org.apache.samza.sql.planner.QueryPlanner;
+import org.apache.samza.sql.runner.SamzaSqlApplicationConfig;
+import org.apache.samza.sql.testutil.SamzaSqlQueryParser;
+
+
+/**
+ * This class is used to populate the StreamGraph using the SQL queries.
+ * This class contains the core of the SamzaSQL control code that converts the SQL statements to calcite relational graph.
+ * It then walks the relational graph and then populates the Samza's {@link StreamGraph} accordingly.
+ */
+public class QueryTranslator {
+
+  private final Map<String, Optional<SamzaToRelMsgConverter>> samzaToRelMsgConvertersBySystem;
+  private final ScanTranslator scanTranslator;
+  private final Map<String, RelSchemaConverter> schemaConvertersBySystem;
+  private final Map<String, SourceSchemaResolver> schemaResolversBySystem;
+  private final Map<String, Optional<RelToSamzaMsgConverter>> relToSamzaMsgConvertersBySystem;
+  private final SamzaSqlApplicationConfig sqlConfig;
+
+  public QueryTranslator(SamzaSqlApplicationConfig sqlConfig) {
+    this.sqlConfig = sqlConfig;
+    schemaResolversBySystem = sqlConfig.getSchemaResolversBySystem();
+    samzaToRelMsgConvertersBySystem = sqlConfig.getSamzaToRelMsgConvertersBySystem();
+    relToSamzaMsgConvertersBySystem = sqlConfig.getRelToSamzaMsgConvertersBySystem();
+    schemaConvertersBySystem = sqlConfig.getSchemaConvertersBySystem();
+    scanTranslator = new ScanTranslator(samzaToRelMsgConvertersBySystem);
+  }
+
+  public void translate(SamzaSqlQueryParser.QueryInfo queryInfo, StreamGraph streamGraph) {
+    QueryPlanner planner = new QueryPlanner(schemaResolversBySystem, schemaConvertersBySystem,
+        sqlConfig.getInputSystemStreamConfigBySource(), sqlConfig.getUdfMetadata());
+    final RelRoot relRoot = planner.plan(queryInfo.getSelectQuery());
```

Comment:
Noticed that in SamzaSqlQueryPlanner.parseQuery(), there is already a SqlSelect generated for the part does the SELECT statement. Following this planner.plan(), in QueryPlanner, we are parsing the select statement again (as queryInfo.getSelectQuery() is returning a string). One question: can we do a single parse instead of double parse?

Reply:
@nickpan47 I think we talked about this in our last sync up meeting. One of the ideas that came out of the meeting is whether we could do one time planning and send the serialized planner output as config to each of the containers. I think that is possible, I will open a Jira for that.